### PR TITLE
Fetch and display ad impressions forecast

### PR DESCRIPTION
**Pull request by JorgeMucientes**
Part of: #10574

### Description

Fetches and displays the ad impressions forecast

- Refactors how state is handle in `BlazeCampaignBudgetViewModel` from live data to a combination of StateFlows
- Fetches impressions forecast numbers after releasing the budget slider. 
- Shows loading state while fetching forecast results. 
- Handle errors and retry when failing to fetch impressions forecast. This matches iOS behavior. 

### Testing instructions

- Trigger Blaze campaign creation
- Open edit budget screen
- Play with budget slider and see how forecast is refreshed
- Remove connectivity and slide budget value, check the error is displayed
- Turn back on the internet connection and click `Retry`

### Images/gif

https://github.com/woocommerce/woocommerce-android/assets/2663464/5a51779c-e2c2-452a-9f4e-d4fbfc3d83a5

- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          2 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">This PR is tagged with <code>status: do not merge</code> label(s).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>ShowDialogFragment</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/BlazeRepository.kt [hunk 1]
```diff
@@ -123,6 +130,27 @@ class BlazeRepository @Inject constructor(
...
 
+    suspend fun fetchAdForecast(
+        startDate: Date,
+        campaignDurationDays: Int,
+        totalBudget: Float
+    ): Result<BlazeAdForecast> {
+        val result = blazeCampaignsStore.fetchBlazeAdForecast(
+            selectedSite.get(),
+            startDate,
+            Date(startDate.time + campaignDurationDays * ONE_DAY_IN_MILLIS),
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/BlazeRepository.kt]: nitpicky np 😅, I find that Kotlin's Duration extensions make these conversions easier to read:
 ```kotlin
             Date(startDate.time + campaignDurationDays.days.inWholeMilliseconds),
```

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt [hunk 2]
```diff
@@ -13,41 +15,57 @@ import com.woocommerce.android.util.CurrencyFormatter
...
 class BlazeCampaignBudgetViewModel @Inject constructor(
     savedStateHandle: SavedStateHandle,
-    private val currencyFormatter: CurrencyFormatter
+    private val currencyFormatter: CurrencyFormatter,
+    private val repository: BlazeRepository
 ) : ScopedViewModel(savedStateHandle) {
     private val navArgs: BlazeCampaignBudgetFragmentArgs by savedStateHandle.navArgs()
 
-    private val _viewState = MutableLiveData(
+    private val campaignForecastState = savedStateHandle.getStateFlow(viewModelScope, getLoadingForecastUi())
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt]: Not sure why we need to have this separate field form the forecast while it's also part of the `BudgetUiState`? with this we end up saving it twice to the saved state.

WDYT about something like the following that changes it to a computed property that delegates changes to the already saved item in `budgetUiState`?
```patch
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt	(revision d589497f970ff719a62f627c54bd9e9c07e4f052)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt	(date 1707499028331)
@@ -18,7 +18,6 @@
 import com.woocommerce.android.viewmodel.getStateFlow
 import com.woocommerce.android.viewmodel.navArgs
 import dagger.hilt.android.lifecycle.HiltViewModel
-import kotlinx.coroutines.flow.combine
 import kotlinx.coroutines.flow.update
 import kotlinx.coroutines.launch
 import kotlinx.parcelize.Parcelize
@@ -33,7 +32,6 @@
 ) : ScopedViewModel(savedStateHandle) {
     private val navArgs: BlazeCampaignBudgetFragmentArgs by savedStateHandle.navArgs()
 
-    private val campaignForecastState = savedStateHandle.getStateFlow(viewModelScope, getLoadingForecastUi())
     private val budgetUiState = savedStateHandle.getStateFlow(
         viewModelScope,
         BudgetUiState(
@@ -56,12 +54,13 @@
         )
     )
 
-    val viewState = combine(
-        campaignForecastState,
-        budgetUiState
-    ) { forecast, budgetUiState ->
-        budgetUiState.copy(forecast = forecast)
-    }.asLiveData()
+    private var campaignForecastState
+        get() = budgetUiState.value.forecast
+        set(value) {
+            budgetUiState.update { it.copy(forecast = value) }
+        }
+
+    val viewState = budgetUiState.asLiveData()
 
     init {
         fetchAdForecast()
@@ -147,28 +146,24 @@
     }
 
     private fun fetchAdForecast() {
-        campaignForecastState.update { it.copy(isLoading = true) }
+        campaignForecastState = campaignForecastState.copy(isLoading = true)
         launch {
             repository.fetchAdForecast(
                 startDate = Date(budgetUiState.value.campaignStartDateMillis),
                 campaignDurationDays = budgetUiState.value.durationInDays,
                 totalBudget = budgetUiState.value.totalBudget
             ).onSuccess { fetchAdForecastResult ->
-                campaignForecastState.update {
-                    it.copy(
-                        isLoading = false,
-                        isError = false,
-                        impressionsMin = fetchAdForecastResult.minImpressions,
-                        impressionsMax = fetchAdForecastResult.maxImpressions
-                    )
-                }
+                campaignForecastState = campaignForecastState.copy(
+                    isLoading = false,
+                    isError = false,
+                    impressionsMin = fetchAdForecastResult.minImpressions,
+                    impressionsMax = fetchAdForecastResult.maxImpressions
+                )
             }.onFailure {
-                campaignForecastState.update {
-                    it.copy(
-                        isLoading = false,
-                        isError = true
-                    )
-                }
+                campaignForecastState = campaignForecastState.copy(
+                    isLoading = false,
+                    isError = true
+                )
             }
         }
     }
```

**Review by hichamboushaba — APPROVED** [review]
Nice work @JorgeMucientes, it works well.

I left some comments, but nothing is blocking.
And I have two non-blocking questions:

1. I see we pass multiple fields in the `navArgs`, why didn't you just pass the `Budget` instance?
2. Same question for the class `EditBudgetAndDurationResult`, why not just reuse the class `Budget` for this too? 

For both of these cases, it seems to me the same fields are present in the class `Budget` and could result in a simpler code, WDYT?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/BlazeRepository.kt (hunk 1 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/BlazeRepository.kt]: Nice! TIL about it. Refactored it: https://github.com/woocommerce/woocommerce-android/pull/10734/commits/0cbf3f9a3f8720e4e5c86a28022d354e42b927e4 and got rid of `ONE_DAY_IN_MILLIS` constant.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt (hunk 2 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt]: Good suggestion! Applied!

**Comment by JorgeMucientes** [comment]
Thanks for the review @hichamboushaba 

> - I see we pass multiple fields in the navArgs, why didn't you just pass the Budget instance?
> - Same question for the class EditBudgetAndDurationResult, why not just reuse the class Budget for this too?

Yeah, good points. I've applied these changes too as it simplifies de code a bit. I'd say one of the reasons I didn't use `Budget` initially was because it lived in the `Repository` and as I shared here: https://github.com/woocommerce/woocommerce-android/pull/10732#discussion_r1484514621 my initial thought about models defined in the repository is that they were business models, so it didn't feel right to use them to pass data around screens. However, upon further clarification: https://github.com/woocommerce/woocommerce-android/pull/10732#discussion_r1484557064, I think it's fine to reuse the `Budget` model and simplify the code. 

Thanks again for the feedback!

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**